File: src/tree.py
from .models import Task, TaskStatus

STATUS_ICONS = {
    TaskStatus.PENDING: "○",
    TaskStatus.IN_PROGRESS: "◈",
    TaskStatus.VERIFYING: "◇",
    TaskStatus.DONE: "◉",
    TaskStatus.FAILED: "✗",
}

STATUS_LABELS = {
    TaskStatus.PENDING: "pending",
    TaskStatus.IN_PROGRESS: "in_progress",
    TaskStatus.VERIFYING: "verifying",
    TaskStatus.DONE: "done ✓",
    TaskStatus.FAILED: "failed",
}


def build_tree(tasks: list[Task]) -> list[Task]:
    """Build tree structure from flat task list.

    Returns root nodes with children populated.
    """
    by_id: dict[str, Task] = {}
    for t in tasks:
        t.children = []
        by_id[t.id] = t

    roots: list[Task] = []
    for t in tasks:
        if t.parent_id and t.parent_id in by_id:
            by_id[t.parent_id].children.append(t)
        else:
            roots.append(t)

    return roots


def render_task(
    task: Task, prefix: str = "", is_last: bool = True, is_root: bool = True
) -> str:
    """Render a single task and its children as a tree string."""
    icon = STATUS_ICONS[task.status]
    label = STATUS_LABELS[task.status]

    if is_root:
        connector = ""
        child_prefix = ""
    else:
        connector = "└── " if is_last else "├── "
        child_prefix = "    " if is_last else "│   "

    line = f"{prefix}{connector}{icon} {task.name} [{label}]"
    lines = [line]

    for i, child in enumerate(task.children):
        is_child_last = i == len(task.children) - 1
        lines.append(
            render_task(child, prefix + child_prefix, is_child_last, is_root=False)
        )

    return "\n".join(lines)


def render_tree(tasks: list[Task]) -> str:
    """Render full task tree from flat list."""
    if not tasks:
        return "No tasks found."

    roots = build_tree(tasks)
    rendered = []
    for root in roots:
        rendered.append(render_task(root))

    return "\n\n".join(rendered)


def render_subtree(task: Task, children: list[Task]) -> str:
    """Render a single task with its direct children."""
    task.children = children
    return render_task(task)

File: internal/domain/tree.go
package domain

import "strings"

// BuildTree builds a tree structure from a flat task list.
// Returns root nodes with Children populated.
func BuildTree(tasks []*Task) []*Task {
	byID := make(map[string]*Task, len(tasks))
	for _, t := range tasks {
		t.Children = nil
		byID[t.ID] = t
	}

	var roots []*Task
	for _, t := range tasks {
		if t.ParentID != nil && byID[*t.ParentID] != nil {
			parent := byID[*t.ParentID]
			parent.Children = append(parent.Children, t)
		} else {
			roots = append(roots, t)
		}
	}
	return roots
}

// renderTask renders a single task and its children as a tree string.
func renderTask(task *Task, prefix string, isLast bool, isRoot bool) string {
	icon := task.Icon()
	label := task.Label()

	var connector, childPrefix string
	if isRoot {
		connector = ""
		childPrefix = ""
	} else if isLast {
		connector = "└── "
		childPrefix = "    "
	} else {
		connector = "├── "
		childPrefix = "│   "
	}

	line := prefix + connector + icon + " " + task.Name + " [" + label + "]"
	lines := []string{line}

	for i, child := range task.Children {
		isChildLast := i == len(task.Children)-1
		lines = append(lines, renderTask(child, prefix+childPrefix, isChildLast, false))
	}

	return strings.Join(lines, "\n")
}

// RenderTree renders a full task tree from a flat list.
func RenderTree(tasks []*Task) string {
	if len(tasks) == 0 {
		return "No tasks found."
	}

	roots := BuildTree(tasks)
	parts := make([]string, 0, len(roots))
	for _, root := range roots {
		parts = append(parts, renderTask(root, "", true, true))
	}
	return strings.Join(parts, "\n\n")
}

// RenderSubtree renders a single task with its direct children.
func RenderSubtree(task *Task, children []*Task) string {
	task.Children = children
	return renderTask(task, "", true, true)
}
